deeplearning4j-nearestneighbors-parent/nearestneighbor-core/src/main/java/org/deeplearning4j/clustering/lsh/RandomProjectionLSH.java

```diff
@@ -0,0 +1,230 @@
+     *
+     * An efficient method for generating uniformly distributed points on the surface of an n-dimensional sphere
+     * JS Hicks, RF Wheeling - Communications of the ACM, 1959
+     * @param data a query to generate multiple probes for
+     * @return `numTables`
+     */
+    public INDArray entropy(INDArray data){
+
+        INDArray data2 =
+                    Nd4j.getExecutioner().exec(new GaussianDistribution(Nd4j.create(numTables, inDimension), radius));
+
+        INDArray norms = Nd4j.norm2(data2.dup(), -1);
+
+        assert(norms.shape()[0] == numTables);
+        assert(norms.shape()[1] == 1);
+
+        data2.diviColumnVector(norms);
+        data2.addiRowVector(data);
+        return data2;
+    }
+
+    /**
+     * Returns hash values for a particular query
+     * @param data a query vector
+     * @return its hashed value
+     */
+    public INDArray hash(INDArray data) {
+        if (data.shape()[1] != inDimension){
+            throw new ND4JIllegalStateException(
+                    String.format("Invalid shape: Requested INDArray shape %s, this table expects dimension %d",
+                            Arrays.toString(data.shape()), inDimension));
+        }
+        INDArray projected = data.mmul(randomProjection);
+        INDArray res = Nd4j.getExecutioner().execAndReturn(new Sign(projected));
+        return res;
+    }
+
+    /**
+     * Populates the index. Beware, not incremental, any further call replaces the index instead of adding to it.
+     * @param data the vectors to index
+     */
+    @Override
+    public void makeIndex(INDArray data) {
+        index = hash(data);
+        indexData = data;
+    }
+
+    // data elements in the same bucket as the query, without entropy
+    INDArray rawBucketOf(INDArray query){
+        INDArray pattern = hash(query);
+
+        INDArray res = Nd4j.zeros(index.shape());
+        Nd4j.getExecutioner().exec(new BroadcastEqualTo(index, pattern, res, -1));
+        return res.min(-1);
+    }
+
+    @Override
+    public INDArray bucket(INDArray query) {
+        INDArray queryRes = rawBucketOf(query);
+
+        if(numTables > 1) {
+            INDArray entropyQueries = entropy(query);
+
+            // loop, addi + conditionalreplace -> poor man's OR function
+            for (int i = 0; i < numTables; i++) {
+                INDArray row = entropyQueries.getRow(i);
+                queryRes.addi(rawBucketOf(row));
+            }
+            BooleanIndexing.replaceWhere(queryRes, 1.0, Conditions.greaterThan(0.0));
+        }
+
+        return queryRes;
+    }
+
+    // data elements in the same entropy bucket as the query,
+    INDArray bucketData(INDArray query){
+        INDArray mask = bucket(query);
+        int nRes = mask.sum(0).getInt(0);
+        INDArray res = Nd4j.create(new int[] {nRes, inDimension});
+        int j = 0;
+        for (int i = 0; i < nRes; i++){
```

Comment:
@raver119 could probably do this in c++ in another rev..

Reply:
Yes, we can fit anything we want into CustomOp...